---
title: "inbreedR step by step"
author: "Martin A. Stoffel"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{inbreedR step by step}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>", cache = FALSE) # warning = FALSE
library(inbreedR)
library(knitr)
```

This vignette provides a short introduction to the functions of the `inbreedR` package. The package in its current form is a minimal version that contains the commonly used and essential functions for the analysis of inbreeding based on microsatellite or snp markers, but potential extensions will follow and any improvement suggestions (from bugs to potential new functions) are always welcome. To download the development version from GitHub just type:

```{r, eval = FALSE}
install.packages("devtools")
library("devtools")
install_github("mastoffel/inbreedR")
library("inbreedR")
```

To get an overview over all the functions just use the help files.

```{r, eval = FALSE}
?inbreedR
```

## Data format

The data format that all functions use is fairly simple. The input always is a `data.frame` or `matrix`, whereby rows represent individuals, and each locus is one column. Heterozygote loci are coded as `1` and homozygote loci as `0`. Missing data should be encoded as NA (numbers other than 0 or 1 are also ok). `mouse_snps` is a sample dataset accompanying the package. 

```{r}
data("mouse_snps")
mouse_snps[1:10, 1:10]
```

You can check whether your data is in the right format with the `check_data` function, which gives an error with a message when something
went wrong and `TRUE` otherwise. Look at `?check_data` to see what exactly is checked for.

```{r}
check_data(mouse_snps, num_ind = 36, num_loci = 13198)
```

### Conversion from a more common format

`convert_raw` is a function to convert a more common format, where each locus is represented by two coloumns (alleles), into the inbreedR working format. Microsatellite data are often formatted like this:

```{r}
data("mouse_msats")
mouse_msats[1:8, 1:8]
```

The data.frame itself should just contain the marker data, while rownames for individuals or columnnames for loci are possible. To convert it into the `inbreedR` working format, just use the `convert_raw` function, specify the missing value and assign it to a new variable.

```{r}
genotypes <- convert_raw(mouse_msats, miss = NA) 
```

It is now converted to the working format, while missings are coded as -1. The same procedure works when
you have letters (e.g. basepairs "A", "T") in two adjacent columns instead of microsatellite length numbers.

## Identity disequilibrium with g2 and HHC

Two different methods have been used to assess the correlation of heterozygosity across loci, known as **identity disequilibrium (ID)**, which represents variation in inbreeding. **ID** thus builds the foundation for so called global or general effects of heterozygosity on fitness.

### g2

**g2** measures the excess of double heterozygote loci in a population. The original formula by [@david2007reliable] is implemented in the `g2_microsats` function, which also allows calculation of a confidence interval based on bootstrapping (nboot) and a p-value based on permutation tests (nperm).

```{r, results = "hide"}
g2_seals <- g2_microsats(genotypes, nperm = 100, nboot = 100, CI = 0.95)
```

Just printing the object gives a summary on the results.

```{r}
g2_seals
```

Plotting the object gives a distribution of the bootstrapped values and a confidence interval.

```{r,fig.width = 6, fig.height = 5}
plot(g2_seals)
```

The plots are also customisable to a certain degree. Often you may like to have a more fine or wide scaled
bar pattern or different labels. Just pass any argument that you can also pass to the `hist()` function. See `?hist`.

```{r,fig.width = 6, fig.height = 5}
plot(g2_seals, main = "just another title", xlab = "identity disequilibrium (g2)", breaks = 5, col = "red")
```

Calculating g2 based on the original formula is computationally expensive and unpractical for large (i.e. SNP) datasets, especially where bootstraps/permutations are required, as this requires double summation over all pairs of loci. [@hoffman2014high] present a computationally more tractable form, based on the assumption that the number of missing values does not vary a lot among individuals.

```{r, results = "hide"}
g2_mouse_snps <- g2_snps(mouse_snps, nperm = 10, nboot = 10, CI = 0.95)
```

Showing results and plotting as before.

```{r,fig.width = 6, fig.height = 5}
g2_mouse_snps
plot(g2_mouse_snps)
```

You might want to increase permutations and bootstraps for calculating p-values and CIs. For future versions of the package, we are planning to allow parallel computations for further speeding up the process.

### Heterozygosity-heterozygosity correlations (HHC)

HHC is another way to quantify inbreeding. The marker data is randomly split into two equal parts and the correlation between the resulting standardized multilocus heterozygosity (sMLH) values is calculated. This is repeated over n iterations. A distribution of values that differ significantly from zero provides evidence for a correlation in heterozygosity across loci that can be attributed to variation in the inbreeding coefficients of individuals present in the sample. Running this for the microsatellite dataset:

```{r, results = "hide"}
hhc_microsats <- HHC(genotypes, niter = 100, CI = 0.95)
```

As before, we can print the summary... 

```{r}
hhc_microsats
```

...and plot the distribution.

```{r,fig.width = 6, fig.height = 5}
plot(hhc_microsats)
```

## Sensitivity to the number of markers 

### g2
One question adressed by recent papers [@miller2014assessment, @hoffman2014high] is the sensitivity of g2 towards the number of markers used.
The function `resample_g2()` calculates g2 for different marker subsets specified by the user.

```{r, results = "hide"}
g2_devel <- resample_g2(genotypes, subsets = c(2,4,6,8,10,12), nboot = 1000, type = "msats")
```

```{r, results = "hide"}
g2_devel
```

A typical way to show the g2 distributions for different subsets of markers are boxplots. `inbreedR` provides a simple means of plotting the results.

```{r,fig.width = 5, fig.height = 4}
plot(g2_devel)
```

You can also customise most elements of these boxplots. Just add any arguments that you would usually pass to `boxplot()`. See `?boxplot()` for infos. 

```{r,fig.width = 6, fig.height = 5}
plot(g2_devel, main = "identity disequilibrium for different marker subsets", col = "green")
```


### Expected $r^2$ between heterozygosity and inbreeding

When looking at heterozygosity-fitness correlations (HFCs), one of the central questions is how well heterozygosity (sMLH) represents the level of inbreeding. According to [@szulkin2010heterozygosity], the expected correlation between sMLH and inbreeding level (f) is $$r^2(sMLH, f) = \frac{g2}{\sigma^2(sMLH)}$$

To explore sensitivity to the number of markers towards, we use the function `exp_r2`. This calculates the expected $r^2$ between f and sMLH for different sized, randomly selected, subsets of loci.  The user needs to specify the subset sizes for the analysis.  For example, for a dataset of 10,000 SNPs, one could specify `subsets = c(1, 10, 100, 1000, 2000, 3000, 4000, 5000, 10000)`.  In addition the user should specify how often the current subset is drawn from the overall marker set to calculate the expected r2, and the type of marker, which determines the g2 function to be used.

```{r, results = "hide"}
result <- exp_r2(genotypes, subsets = c(2,4,6,8,10,12), nboot = 100, type = "msats")
```

Then again you can look at the summarized results...

```{r}
result
```

And plot the distribution of expected r2s for each subset.
```{r,fig.width = 6, fig.height = 5}
plot(result)
```

Again, you can customise certain aspects of the plot by adding `boxplot()` arguments. Next to changing labels and the title, you may find notched boxplots quite appealing.

```{r,fig.width = 6, fig.height = 5}
plot(result, notch = TRUE, col = "purple")
```

## HFCs

Last but not least, the package provides a computationally optimized function for the calculation of standardized multilocus heterozygosity (sMLH), which can be used within bootstrapping and permutation tests also for large datasets.

```{r}
dim(mouse_snps) # 13198 loci
system.time(sMLH(mouse_snps))
```

`sMLH` gives a simple vector output with each individuals sMLH.
```{r,fig.width = 6, fig.height = 5}
het <- sMLH(mouse_snps)
hist(het)
```

## Extracting raw data from inbreedR objects 

You may wish to extract and plot the data yourself. Most function outputs are `inbreed` objects and lists. In the `Value` section of each functions documentation (`?fun` ), you can see the data which you can extract. Alternatively, use `str()` to look at the object's structure. Just index the function output with `[["."]]` or `$` as in the following example:

Running the function.

```{r, results = "hide"}
g2_seals <- g2_microsats(genotypes, nperm = 100, nboot = 100, CI = 0.95)
``` 

Looking at the structure.

```{r}
str(g2_seals)
```

Now extract whatever you want from the object, such as the g2 bootstrap results.

```{r}
g2_bootstrap_results <- g2_seals$g2_boot
str(g2_bootstrap_results)
```


## Literature